solidity/test/integration/Slashing.test.ts

```diff
@@ -0,0 +1,529 @@
+/* eslint-disable no-await-in-loop */
+/* eslint-disable @typescript-eslint/no-extra-semi */
+import { ethers, helpers, waffle } from "hardhat"
+import { expect } from "chai"
+
+import type { FakeContract } from "@defi-wonderland/smock"
+import type {
+  ContractTransaction,
+  Contract,
+  BigNumberish,
+  BytesLike,
+} from "ethers"
+import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+
+import type {
+  TBTC,
+  Bridge,
+  TBTCVault,
+  TestRelay,
+  IRandomBeacon,
+  WalletRegistry,
+} from "../../typechain"
+
+import {
+  Operators,
+  performEcdsaDkg,
+  produceOperatorInactivityClaim,
+  updateWalletRegistryDkgResultChallengePeriodLength,
+} from "./utils/ecdsa-wallet-registry"
+import { produceRelayEntry } from "./utils/random-beacon"
+
+import { assertGasUsed } from "./utils/gas"
+import { fixture } from "./utils/fixture"
+
+import {
+  wallet as fraudulentWallet,
+  nonWitnessSignSingleInputTx,
+} from "../data/fraud"
+import { walletState } from "../fixtures"
+import { SingleP2SHDeposit, NO_MAIN_UTXO } from "../data/deposit-sweep"
+import { UTXOStruct } from "../../typechain/Bridge"
+
+const { increaseTime } = helpers.time
+const { createSnapshot, restoreSnapshot } = helpers.snapshot
+
+const describeFn =
+  process.env.NODE_ENV === "integration-test" ? describe : describe.skip
+
+describeFn("Integration Test - Slashing", async () => {
+  let tbtc: TBTC
+  let bridge: Bridge
+  let tbtcVault: TBTCVault
+  let staking: Contract
+  let walletRegistry: WalletRegistry
+  let randomBeacon: FakeContract<IRandomBeacon>
+  let relay: FakeContract<TestRelay>
+  let deployer: SignerWithAddress
+  let governance: SignerWithAddress
+  let thirdParty: SignerWithAddress
+
+  const dkgResultChallengePeriodLength = 10
+
+  before(async () => {
+    ;({
+      deployer,
+      governance,
+      tbtc,
+      bridge,
+      tbtcVault,
+      staking,
+      walletRegistry,
+      relay,
+      randomBeacon,
+    } = await waffle.loadFixture(fixture))
+    ;[thirdParty] = await helpers.signers.getUnnamedSigners()
+
+    // Update only the parameters that are crucial for this test.
+    await updateWalletRegistryDkgResultChallengePeriodLength(
+      walletRegistry,
+      governance,
+      dkgResultChallengePeriodLength
+    )
+  })
+
+  describe("notifyFraudChallengeDefeatTimeout", async () => {
```

Comment:
There are four scenarios where slashing happens:
- `notifyFraudChallengeDefeatTimeout`
- `notifyRedemptionTimeout`
- `notifyMovingFundsTimeout`
- `notifyMovedFundsSweepTimeout`

We have integration tests for the first three.
In theory we could add the test for `notifyMovedFundsSweepTimeout`, although it probably will be the most difficult (there are two wallets involved).

@nkuba @pdyraga @lukasz-zimnoch 
Do you think we should add such a test, or is it an overkill?

Slashing on `notifyMovedFundsSweepTimeout` should  use about the same amount of gas as in other cases.

Reply:
Let's merge it please and consider adding the 4th scenario in another PR